### PR TITLE
Extend Travis build into a matrix pipeline with multiple jobs

### PR DESCRIPTION
**Pull request by danielgustafsson**
This extends our coverage by defining a build matrix in the Travis CI pipeline rather than just having a single job. The idea is to build PRs against combinations of common platforms and compilers in order to catch compiler errors and warnings early. The gist of the changes to the previous CI configuration is:

  - static configuration is replaced by per build config in a matrix
  - ccache is removed to avoid caching issues
  - macOS is again built on
  - the version printing is more selective to only print versions
    for binaries built in every job
  - notifications are removed to minimize spamming
  - builds with various un-orthodox flags are supported
  - build silently to make output more manageable

The second commit is a cherry-pick from upstream to fix the AC_CHECK_DECLS macro on clang 7 and 8.

To see this in action, click the Travis check link in the Checks window below. In there you will instead of the usual buildlog see 4 individual jobs for bionic/gcc, xenial/clang, macos/xcode11 and a one build without any compression algorithms (to catch bugs by assumptions).

Future tasks is to extend this to more combinations and also to build clientside tooling on Windows.

Update: the new version of this PR now also runs a subset of ICW (the upstream tests). The reason for it just being a subset it because Travis is quite picky about output being generated, and some of the slow GPDB tests make builds time out since Travis thinks they've hung. This does not give full coverage of course, but it does significantly increase the coverage in Travis over the current status.

### CODE REVIEW

**Comment by danielgustafsson** [comment]
This is a resubmission of #8258 which was the WIP PR that I used to test and tweak this. That PR is still open since I have a few small things I'd still like to try, but if they work out they will be for a follow-up PR so thats why this is submitted now.

**Comment by gkokolatos** [comment]
This is great!

I have a question though. [Job #30782.1](https://travis-ci.org/greenplum-db/gpdb/jobs/575051252) of this PR sets `CC=gcc-8` yet
> $ gcc --version
gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
Copyright (C) 2017 Free Software Foundation, Inc.

So further down when I it shows lines like
> gcc -DNDEBUG -g -fwrapv ....

How one can know that it actually used gcc-8? Or did it use the preinstalled gcc 7.4.0 of the OS?

**Comment by danielgustafsson** [comment]
> How one can know that it actually used gcc-8? Or did it use the preinstalled gcc 7.4.0 of the OS?

This was a bug, executing the overrides got lost in a rebase. Fixed now.

File: .travis.yml [hunk 1]
```diff
@@ -1,108 +1,183 @@
...
+            --with-libxml \
+            --enable-mapreduce \
+            --enable-orafce \
+            $C
+        make -s install
+        source ${TRAVIS_BUILD_DIR}/gpsql/greenplum_path.sh
+        make -s unittest-check
+        make -C gpAux/gpdemo cluster
+        source gpAux/gpdemo/gpdemo-env.sh
+        make -C src/test/regress installcheck-small
```
Comment (asimrp) [inline .travis.yml]: Not in this patch but at some point in future, we may want to create a different target to define tests that can run against a production build, which should exclude any tests using fault injector.

File: .travis.yml [hunk 2]
```diff
@@ -1,108 +1,183 @@
...
+        #
+        # Configuration variations
+        # ----------------------------------------------------------------
+        #
+        # Debug build without any compression algorithms supplied
+        - os: linux
+          dist: bionic
+          compiler: gcc
+          env:
+              - T=debug C="--without-zlib --without-libbz2 --without-zstd --without-quicklz"
```
Comment (asimrp) [inline .travis.yml]: This is a very thoughtful configuration, thank you for addressing that.  To make it more effective, the set of tests that is run against this configuration should be able to exercise code paths that make use of compression algorithms.  E.g. appendonly tests or the tests that create spill files (I'm guessing `libbz2` is used for spill files).

File: .travis.yml (hunk 1 above)
Comment (danielgustafsson) [reply .travis.yml]: Agreed. Ideally we'd reach the point where we don't require `plpython` for regression tests too such that we can do `--without-python` builds, but thats yet again another thing.

File: .travis.yml (hunk 2 above)
Comment (danielgustafsson) [reply .travis.yml]: bz2 used by gfile, which in turn is used by gpfdist and gpload among others.

The reason I made this configuration was simply that we had broken it badly, so it seemed as  good an example as any. I agree that there should be tests for this, and simply running the usual regression tests with alternative output files for non-encryption-builds is one way. Not sure what would be the best option.

**Comment by danielgustafsson** [comment]
> Quick question - how to interpret the two times reported on the Travis page?
> 
> > Ran for 16 min 9 sec
> > Total time 53 min 14 sec

The total wallclock time for running the jobs in the pipeline in parallel was 16 minutes and 9 seconds, but the actual cputime spent for running these jobs were 53 minutes and 14 seconds.

As a comparison, the [curl](https://github.com/curl/curl) Travis pipeline runs almost 30 builds in ~ 3 hours with over 10 cpuhours spent. The longest job is however only about 40 minutes, but since the jobs aren't started at the same time, the overall runtime can be significantly longer than the longest job (especially when running many jobs).

**Comment by asimrp** [comment]
~ 16 minutes to get feedback about several platforms with a few tests on each platform seems quite impressive and desirable.  PR pipeline takes 1 hour 30 minutes, some times 3+ hours to give feedback for a single platform, depending on load on the Concourse infrastructure.

**Comment by danielgustafsson** [comment]
Pushed to master, thanks for review!

**Comment by d** [comment]
> ~ 16 minutes to get feedback about several platforms with a few tests on each platform seems quite impressive and desirable. PR pipeline takes 1 hour 30 minutes, some times 3+ hours to give feedback for a single platform, depending on load on the Concourse infrastructure.

@asimrp you are comparing apples to oranges. For reference, the greenplum_schedule dominates the runtime of ICG (parallel_schedule + greenplum_schedule). Also Concourse actually provides *more* compute resource to builds on average (8 vCPU)

**Pull request closed without merging**